### ISSUE TITLE
RFC: API to run DDlog instance in a separate process.

### ISSUE THREAD

**Issue by ryzhyk** [vmware/differential-datalog#794]
Motivation
------------

- Run multiple instances of the same program on behalf of multiple tenants while enforcing resource isolation between tenants.
- Run different DDlog programs for the same or multiple tenants (DDlog-generated '.so' libraries export the same set of symbols; therefore multiple libraries cannot be easily loaded in the same process).

Design
---------

- This will probably be implemented using the D3log infrastructure.
- We may even want to support remote DDlog instances.

**details TBD**

**Comment by Kixiron** [comment]
It may be awkward in the actual implementation, but I think basing the api off of an in-process model is the best route for having an api that feels good to work with. Ideally you could totally swap backends from ddlog to d3log with little to no changes in code, even though that's a slightly unrealistic goal to have

**Comment by mbudiu-vmw** [comment]
>  statically typed API will require the client to additionally link against a DDlog-generated library.
Do you intend to support the case of client that loads multiple such libraries at once?

**Comment by mbudiu-vmw** [comment]
> Starting a DDlog instance.

Maybe use ssh for this? Are there any requirements for the environment of a ddlog process? I can't really think of any myself - if you can run a rust executable you should be able to run ddlog.

**Comment by mbudiu-vmw** [comment]
> Split trait DDlog into statically and dynamically typed APIs.

You should probably start with this piece, since it has nothing to do with the distributed execution.

**Comment by mbudiu-vmw** [comment]
> The DDlog protocol, implemented over the ZeroMQ protocol

ZeroMQ seems to offer much more than you need. For example, I think you will never need multicast.

> heartbeat messages

If you use an existing framework it should take care of this. You don't want to implement this by hand.

**Comment by ryzhyk** [comment]
> > statically typed API will require the client to additionally link against a DDlog-generated library.
> 
> Do you intend to support the case of client that loads multiple such libraries at once?

Not at the moment, but this is in any case orthogonal to this RFC.

**Comment by mbudiu-vmw** [comment]
How about having two separate traits for producers (who supply deltas) and consumers (who implement the callbacks)? This way the consumers don't have to be in the same process and you can make a chain of processes.

**Comment by ryzhyk** [comment]
> > Starting a DDlog instance.
> 
> Maybe use ssh for this? Are there any requirements for the environment of a ddlog process? I can't really think of any myself - if you can run a rust executable you should be able to run ddlog.

Right, this would be one option. One could also use some form of container technology, a dedicated factory server, or any other method of running a program. I don't think we need to prescribe anything here.

**Comment by ryzhyk** [comment]
> > Split trait DDlog into statically and dynamically typed APIs.
> 
> You should probably start with this piece, since it has nothing to do with the distributed execution.

yep

**Comment by ryzhyk** [comment]
> > The DDlog protocol, implemented over the ZeroMQ protocol
> 
> ZeroMQ seems to offer much more than you need. For example, I think you will never need multicast.

It does, but I don't need to use all of ZeroMQ, and there are many useful parts I actually want to use.

> > heartbeat messages
> 
> If you use an existing framework it should take care of this. You don't want to implement this by hand.

ZeroMQ makes it relatively easy to do this, but I don't think there's a built in primitive.

**Comment by mbudiu-vmw** [comment]
You will have to think about the semantics of the various primitives. What does the return result from apply_updates mean? That the updates have been queued, received by the server, or that the computation is completed and changes have been already sent to the callback?

**Comment by ryzhyk** [comment]
> You will have to think about the semantics of the various primitives. What does the return result from apply_updates mean? That the updates have been queued, received by the server, 

it means that updates have been queued somewhere. It does not matter where as long as they will be applied in the next commit. The only guarantee we provide is that, if the commit succeeded then all preceding updates have been applied in order (i.e., they don't get lost or reordered). If DDlog cannot enforce this, it will go into an error state. 

> or that the computation is completed and changes have been already sent to the callback?

`apply_updates` never meant that.

**Comment by ryzhyk** [comment]
> How about having two separate traits for producers (who supply deltas) and consumers (who implement the callbacks)? This way the consumers don't have to be in the same process and you can make a chain of processes.

Definitely, not another producer/consumer pattern :) You can totally chain things using `trait DDlog` (or whatever we evolve it into).
